server, storage: remove Open() from the Engine interface

andreimatei opened the pull request:
This is a series of refactorings whose main points are to remove the distinction between creating and opening an Engine, and then moving the creation of the Engines inside `Server.Start()`.
They were done for #9504, which needs to read from the engines inside `Server.Start()`, before the engines used to be `Open()ed`.

CC @BramGruneir for r3, @RaduBerinde for r2 which moves `noCopy` to a central location.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9845)

<!-- Reviewable:end -->

andreimatei committed:
storage: remove Engine.Open()

Prior to this commit, engine initialization and Open()ing were separate
operations, mirroring the underlying interface provided by RocksDB.
However keeping these operations separate suggest it is possible to open
an engine multiple times, which we do not intend to support (it was
partly suported only for memory-based unit tests, but you couldn't
re-Open() an engine after Close()). So, instead, we combine
initialization with open, so that an Engine is always in the opened
state and this subtlety removed.

We also move towards explicit ownership of the engines by no longer
having the factory schedule their destruction, and letting the caller do
it.

andreimatei committed:
server: take engines out or Server.Context

They didn't really belong there, as the Context is supposed to hold
configuration, not this kind of objects. Moreover, the fact that they
were hanging from that Context made their ownership awkward.
Now the Server takes ownership.

andreimatei committed:
cli, storage: remove engine snapshotting debug feature

It's not used, races with engine stops, and uses the engines outside of the Server,
which is a problem for the next commits.

fixes #9815

andreimatei committed:
server: move creation of Engines in Server.Start

It was unnecessary for it to happen earlier and it meant an extra
transfer of ownership.

Now that Engines.Move() is not used in function calls, we change it
to return by value.

andreimatei committed:
server: unify the way engines are created

... between tests and prod - handle everything in
ServerContext.InitStores(). So that a future commit can move that to the
Server.Start().

RaduBerinde commented:
I think @knz did the same with NoCopy in #9821 (10d510aa). Definitely good to have it in a common place.

petermattis commented:
Reviewed 1 of 1 files at r1, 29 of 29 files at r2, 4 of 4 files at r3, 6 of 6 files at r4.
Review status: 26 of 31 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[server/context.go, line 365 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjP6Mynz6MtJ-5jpxg:-KTjP6Mynz6MtJ-5jpxh:b-dujzxf) ([raw file](https://github.com/cockroachdb/cockroach/blob/defe6f35ee104729f2a164b1c69d7cd0de246421/server/context.go#L365)):*
> ```Go
> // 	ep.Release()
> //
> type EnginesUniquePtr struct {
> ```

This naming is very C++-esque. Not a fan because this struct doesn't model a pointer like `unique_ptr` does in C++. Perhaps `EnginesContainer`?

---

*[server/context.go, line 468 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjPjcLFojWMjjQ-pnK:-KTjPjcLFojWMjjQ-pnL:b-xt1lsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/defe6f35ee104729f2a164b1c69d7cd0de246421/server/context.go#L468)):*
> ```Go
> 				ctx.Engines,
> 				eng,
> 			)
> ```

This could be a single line now (unless you're trying to bump up your klocs stats).

---

*[server/context.go, line 396 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjyXrpTTwo8E8AGim8:-KTjyXrpTTwo8E8AGim9:b29bo5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/86dbef64edfac81f0ee0f9bd7b5df9b7b51f9633/server/context.go#L396)):*
> ```Go
> // Move transfers ownership of the engines another EnginesUniquePtr.
> // A future call to Close() on the original unique ptr will be a no-op.
> func (ep *EnginesUniquePtr) Move() *EnginesUniquePtr {
> ```

`Move` is also a C++-ism. I preferred `Release`.

---

*[server/context.go, line 413 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjz0JmDZv0QBjWTnMO:-KTjz0JmDZv0QBjWTnMP:b-kxnwzf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8733ef5cabdbe419a4dd85a942466128d5c618d/server/context.go#L413)):*
> ```Go
> 
> // CreateEngines creates Engines based on the specs in ctx.Stores.
> func (ctx *Context) CreateEngines() (engines *EnginesUniquePtr, err error) {
> ```

I'd nix the named return values here. You don't need them as you're always explicitly returning and you're not trying to manipulate one of the return values in a `defer`.

---

*[server/context.go, line 413 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjzTPAjRC95YR3bo03:-KTjzTPAjRC95YR3bo04:b-14vsw8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8733ef5cabdbe419a4dd85a942466128d5c618d/server/context.go#L413)):*
> ```Go
> 
> // CreateEngines creates Engines based on the specs in ctx.Stores.
> func (ctx *Context) CreateEngines() (engines *EnginesUniquePtr, err error) {
> ```

We're using the verb `Init` in this code. Why did you switch to `Create`? `InitEngines` seems good.

---

*[storage/engine/rocksdb.go, line 370 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjQUFl1r4h1la43GeM:-KTjQUFl1r4h1la43GeN:b3ixuao) ([raw file](https://github.com/cockroachdb/cockroach/blob/defe6f35ee104729f2a164b1c69d7cd0de246421/storage/engine/rocksdb.go#L370)):*
> ```Go
> 	}
> 	return r, nil
> }
> ```

Unnecessary movement of this function makes the commit harder to review. Mind moving it back?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
TFTR

---

Review status: 26 of 31 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[server/context.go, line 365 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjP6Mynz6MtJ-5jpxg:-KTk0wMQxxOYyJ7nTVnJ:b-mehaek) ([raw file](https://github.com/cockroachdb/cockroach/blob/defe6f35ee104729f2a164b1c69d7cd0de246421/server/context.go#L365)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This naming is very C++-esque. Not a fan because this struct doesn't model a pointer like `unique_ptr` does in C++. Perhaps `EnginesContainer`?
</details>
The C++ motif was intentional. I started with `EnginesHolder` and then moved to this because it's more suggestive of its purpose. The `unique` part is certainly true. `Ptr`... I think this is the closest you can do in go to modeling the `std::unique_ptr`. I've considered embedding the `[]engine.Engine`, but you can't do it just like that - you have to give a new type name to the `[]Engine` - and also I didn't really feel the need for the magic.
Do you really hate it? It is a unique pointer in spirit.

---

*[server/context.go, line 468 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjPjcLFojWMjjQ-pnK:-KTkEb6mUjzSjh6bQlRt:b-vzpeu9) ([raw file](https://github.com/cockroachdb/cockroach/blob/defe6f35ee104729f2a164b1c69d7cd0de246421/server/context.go#L468)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This could be a single line now (unless you're trying to bump up your klocs stats).
</details>
Done, but I'm hoping the klocs stay the same.

---

*[server/context.go, line 396 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjyXrpTTwo8E8AGim8:-KTk1TRaineVP9dB7DNh:bvxwbh3) ([raw file](https://github.com/cockroachdb/cockroach/blob/86dbef64edfac81f0ee0f9bd7b5df9b7b51f9633/server/context.go#L396)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `Move` is also a C++-ism. I preferred `Release`.
</details>
This was on purpose. The ownership used to be `Release`d until this commit, now the ownership is `Move`d.

---

*[server/context.go, line 413 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjzTPAjRC95YR3bo03:-KTk6W0NzJp9-UCH6CP-:b-tic068) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8733ef5cabdbe419a4dd85a942466128d5c618d/server/context.go#L413)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> We're using the verb `Init` in this code. Why did you switch to `Create`? `InitEngines` seems good.
</details>
`Init` as used in this struct refers to initializing some of the struct members. Whereas this guy no longer does that, it returns new engines to the callers. Like God intended.

---

*[server/context.go, line 413 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjz0JmDZv0QBjWTnMO:-KTkEh6DM2CZzPVJgMFa:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8733ef5cabdbe419a4dd85a942466128d5c618d/server/context.go#L413)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I'd nix the named return values here. You don't need them as you're always explicitly returning and you're not trying to manipulate one of the return values in a `defer`.
</details>
done

---

*[storage/engine/rocksdb.go, line 370 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjQUFl1r4h1la43GeM:-KTk52e1IgFGSTePmXNs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/defe6f35ee104729f2a164b1c69d7cd0de246421/storage/engine/rocksdb.go#L370)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Unnecessary movement of this function makes the commit harder to review. Mind moving it back?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-:-KTkGC8v-s7ggLYB8MDC:bb73zsl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 28 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[server/context.go, line 365 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjP6Mynz6MtJ-5jpxg:-KUTobZjkt1pnmwSEOOw:b-jqehuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/defe6f35ee104729f2a164b1c69d7cd0de246421/server/context.go#L365)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> The C++ motif was intentional. I started with `EnginesHolder` and then moved to this because it's more suggestive of its purpose. The `unique` part is certainly true. `Ptr`... I think this is the closest you can do in go to modeling the `std::unique_ptr`. I've considered embedding the `[]engine.Engine`, but you can't do it just like that - you have to give a new type name to the `[]Engine` - and also I didn't really feel the need for the magic.
> Do you really hate it? It is a unique pointer in spirit.
</details>
Yeah, I sort of hate it. The type is only pointer-like if I squint so much my eyes are closed. The idiomatic Go naming for a slice of `[]engine.Engine` would be `Engines`. I'd be fine with that name (even though this is a struct containing such a slice).

---

*[server/context.go, line 396 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjyXrpTTwo8E8AGim8:-KUTopWwjVlUApnEyVC5:b-ok3ub7) ([raw file](https://github.com/cockroachdb/cockroach/blob/86dbef64edfac81f0ee0f9bd7b5df9b7b51f9633/server/context.go#L396)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> This was on purpose. The ownership used to be `Release`d until this commit, now the ownership is `Move`d.
</details>
`Detach`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845)*
<!-- Sent from Reviewable.io -->

tamird commented:
First commit message:

> I hear it's not used, racy, and uses the engines outside of the Server,
which is a problem for the next commits.

Let's avoid language like "I hear" - please be specific, and point to the relevant discussion if possible.

Same comment for the second commit message: why is it so conversational? The subject line is "no more engine.Open" and then the first sentence of the body is "removing engine.Open". This is not an essay, it's a commit message; please be succinct and avoid colourful monologue.

---

Reviewed 30 of 30 files at r6, 1 of 26 files at r7.
Review status: 1 of 28 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[server/context.go, line 95 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUXjqEiMf5aqrWi3Wq2:-KUXjqEj23W-dlnnggoc:b3aawfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b66495420766bb04574a99ebda09222c04a8bd65/server/context.go#L95)):*
> ```Go
> 	// meaty objects, as opposed to configuration structs, so they don't belong
> 	// here. Moreover, they need to be Close()d, and it's easy to lose track of
> 	// that fact when they're hid in this Context.
> ```

s/hid/hidden/

---

*[server/context.go, line 350 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUXkKVXi_oVqx2K_5mJ:-KUXkKVXi_oVqx2K_5mK:b-hbvh12) ([raw file](https://github.com/cockroachdb/cockroach/blob/b66495420766bb04574a99ebda09222c04a8bd65/server/context.go#L350)):*
> ```Go
> }
> 
> // EnginesUniquePtr is a holder of engines, offering an easy way to close them,
> ```

Does this construction really help? I don't get it - you're doing a bunch of gymnastics here to emulate ownership, but what does it get you? What compiler-enforced behaviour do you get as a result? It seems to me that it's still trivially easy to circumvent this monster and close things improperly.

You can achieve essentially the same semantics with much more go-idiomatic code:

```
type engines struct {
  ...
  ...
  engines []engine.Engine
}

func (e *engines) Close() {
  for len(e.engines) > 0 {
    eng := e.engines[0]
    e.engines = e.engines[1:]
    eng.Close()
  }
}

func (e *engines) Transfer() engines {
  ret := *e
  e.engines = nil
  return ret
}
```

---

*[server/context.go, line 355 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUXjtITBfs-P2GWFEEH:-KUXjtITBfs-P2GWFEEI:b54miew) ([raw file](https://github.com/cockroachdb/cockroach/blob/b66495420766bb04574a99ebda09222c04a8bd65/server/context.go#L355)):*
> ```Go
> //
> //	engines := CreateEngines()
> // 	ep := NewEnginesUniquePtr(engines)
> ```

alignment is broken here. probably a mix of tabs and spaces.

---

*[server/context.go, line 366 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUXjy-IZ_hpjJC_WItZ:-KUXjy-IZ_hpjJC_WIt_:brty0yc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b66495420766bb04574a99ebda09222c04a8bd65/server/context.go#L366)):*
> ```Go
> //
> type EnginesUniquePtr struct {
> 	// EnginesUniquePtr is not copiable, as that would leave two owning pointers
> ```

copyable

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-:-KUXi8fFBsx-y5ng6FR4:b4bonin)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
For the first commit, "I hear it's not used" expressed uncertainty and invites discussion, if somebody thinks otherwise. It's hard to "be specific" when asserting that nobody is using something. I've added more specificity around the "racy" part though.

For the second commit, your objection is that "no more Engine.Open" is too colorful? So if I'd have said "remove" instead of "no more" that'd be fine? And how important is it really to not repeat these 3 words in the first line of the body? They create a more natural transition to the second sentence, which starts with "Instead,".

Anyway, I think I also disagree fundamentally with the "no color" and no conversation tone. These messages are written by humans for humans. As long as the "color" doesn't hurt readers and makes prose less precise than it would otherwise be, I think it's fine. Or, when there's uncertainty that you want to express in your prose, the choice of tone can also be a valid way to do so.

---

Review status: 1 of 28 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[server/context.go, line 365 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjP6Mynz6MtJ-5jpxg:-KUXqJB-Xqbv6Ic6ahvA:bxxfbs0) ([raw file](https://github.com/cockroachdb/cockroach/blob/defe6f35ee104729f2a164b1c69d7cd0de246421/server/context.go#L365)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yeah, I sort of hate it. The type is only pointer-like if I squint so much my eyes are closed. The idiomatic Go naming for a slice of `[]engine.Engine` would be `Engines`. I'd be fine with that name (even though this is a struct containing such a slice).
</details>
renamed to `Engines` because I liked the squiting expression.

---

*[server/context.go, line 396 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KTjyXrpTTwo8E8AGim8:-KUXqOasN-ckpYhohspc:b-8wnx35) ([raw file](https://github.com/cockroachdb/cockroach/blob/86dbef64edfac81f0ee0f9bd7b5df9b7b51f9633/server/context.go#L396)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `Detach`?
</details>
I think `Move()` is more descriptive.

---

*[server/context.go, line 95 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUXjqEiMf5aqrWi3Wq2:-KUXqoaxZzz1UnaFi1GD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b66495420766bb04574a99ebda09222c04a8bd65/server/context.go#L95)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/hid/hidden/
</details>
Done.

---

*[server/context.go, line 350 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUXkKVXi_oVqx2K_5mJ:-KUXpZUtBCmBbUv6tNmp:b-gkc6z3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b66495420766bb04574a99ebda09222c04a8bd65/server/context.go#L350)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Does this construction really help? I don't get it - you're doing a bunch of gymnastics here to emulate ownership, but what does it get you? What compiler-enforced behaviour do you get as a result? It seems to me that it's still trivially easy to circumvent this monster and close things improperly.
> 
> You can achieve essentially the same semantics with much more go-idiomatic code:
> 
> ```
> type engines struct {
>   ...
>   ...
>   engines []engine.Engine
> }
> 
> func (e *engines) Close() {
>   for len(e.engines) > 0 {
>     eng := e.engines[0]
>     e.engines = e.engines[1:]
>     eng.Close()
>   }
> }
> 
> func (e *engines) Transfer() engines {
>   ret := *e
>   e.engines = nil
>   return ret
> }
> ```
</details>
The idea is that a client can construct one of these `EnginesUniquePtr` (now renamed) and immediately do `defer e.Close()`, and so if ownership hasn't been transferred by then, you're guaranteed to close them.
Not sure how the code here is different from what you're proposing.

Note that `Transfer` is coming in a later commit (now we only have `Release()`; it will transition to `Move`) because the `Server`, which takes ownership of these guys, is not yet prepared to take them directly.

---

*[server/context.go, line 355 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUXjtITBfs-P2GWFEEH:-KUXp-S6ZpEv_c-LNj3Z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b66495420766bb04574a99ebda09222c04a8bd65/server/context.go#L355)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> alignment is broken here. probably a mix of tabs and spaces.
</details>
Done.

---

*[server/context.go, line 366 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUXjy-IZ_hpjJC_WItZ:-KUXqIG87ZaZt_-s9sU2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b66495420766bb04574a99ebda09222c04a8bd65/server/context.go#L366)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> copyable
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-:-KUXqufJZOMKyy25jrBh:beoyxpi)*
<!-- Sent from Reviewable.io -->

tamird commented:
The second commit message is a monologue:
> Is it possible to
> Open() an engine multiple twice? Well, it was, for RocksDB, but that
> didn't appear to be the intent of the interface. Is it OK to
> Open/Close/Open? No.

Nobody else writes commit messages this way, and it is important to have a consistent tone when describing subtleties like this.

Relying on tone to express uncertainty is fragile and unfriendly to e.g. non-native speakers. For example, instead of "i hear", you could say "I believe", which would be less colloquial and more consistent with other developers' practices.

---

Reviewed 2 of 26 files at r7.
Review status: 0 of 28 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[server/context.go, line 350 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUXkKVXi_oVqx2K_5mJ:-KUYIwuJVXjT51_lHNL8:b-pfb3jv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b66495420766bb04574a99ebda09222c04a8bd65/server/context.go#L350)):*
> The idea is that a client can construct one of these EnginesUniquePtr (now renamed) and immediately do defer e.Close(), and so if ownership hasn't been transferred by then, you're guaranteed to close them.
> Not sure how the code here is different from what you're proposing.

The code you've written is just more complicated and tries to look like C++.

I don't understand the comment about the future commit. Can you squash whatever juggling was done here to one commit so that the same code doesn't need to be reviewed N times?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-:-KUYIeHhbRRu3VCFED4v:bvbkmrm)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 26 files at r12, 3 of 6 files at r14, 1 of 3 files at r15.
Review status: 6 of 28 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/server/config.go, line 349 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUYKyTSYmxagrlbGxOL:-KUYKyTSYmxagrlbGxOM:b-9xyuyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fa1da91fffe5d63db89c6d41373cc28fde50c64/pkg/server/config.go#L349)):*
> ```Go
> // The intended use is:
> //
> //  engines := CreateEngines()
> ```

this alignment is still broken.

---

*[pkg/server/config.go, line 383 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUYLB6byqzicPkEulVh:-KUYLB6byqzicPkEulVi:b-fx6ygh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fa1da91fffe5d63db89c6d41373cc28fde50c64/pkg/server/config.go#L383)):*
> ```Go
> // no-op.
> func (ep *Engines) Close() {
> 	if ep.closed {
> ```

instead of this boolean and a panic, you can reslice as i suggested.

---

*[pkg/server/config.go, line 395 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUYLGieqVqcOwYDVoP3:-KUYLGieqVqcOwYDVoP4:b-qx4bjr) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fa1da91fffe5d63db89c6d41373cc28fde50c64/pkg/server/config.go#L395)):*
> ```Go
> // Move transfers ownership of the engines to another Engines.
> // A future call to Close() on the original Engines will be a no-op.
> func (ep *Engines) Move() *Engines {
> ```

there is no need for this function to return a pointer, and the same is true of `NewEngines`. I believe `go vet` won't complain about `NewEngines`, but I am less sure about this one. You might need to work around it with:
```
defer ep.engines = nil
return NewEngines(ep.GetEngines())
```
to appease `vet`.

---

*[pkg/server/config.go, line 414 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUYL_7e-Vh1QRtKfa5s:-KUYL_7e-Vh1QRtKfa5t:b-scj4ts) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fa1da91fffe5d63db89c6d41373cc28fde50c64/pkg/server/config.go#L414)):*
> ```Go
> func (cfg *Config) CreateEngines() (*Engines, error) {
> 	engines := NewEngines(nil)
> 	// If we're returning engines to the called, they will have been Move()d.
> ```

caller

---

*[pkg/server/config.go, line 419 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUYLcDBuXdbJ3FwX7bb:-KUYLcDBuXdbJ3FwX7bc:b-qgy3ky) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fa1da91fffe5d63db89c6d41373cc28fde50c64/pkg/server/config.go#L419)):*
> ```Go
> 
> 	if cfg.enginesCreated {
> 		return &Engines{}, errors.Errorf("engines already created")
> ```

why do all of these error returns allocate?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 6 of 28 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/server/config.go, line 402 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUYWhrP0I06Kbkoaymi:-KUYWhrP0I06Kbkoaymj:b-q15ci6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fa1da91fffe5d63db89c6d41373cc28fde50c64/pkg/server/config.go#L402)):*
> ```Go
> 
> // AddEngine appends one more Engine to the list managed by this container.
> func (ep *Engines) AddEngine(eng engine.Engine) {
> ```

Nit: `s/AddEngine/Add/g`. You're already dealing with engines, the `Engine` suffix is repetitive. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
I like that monologue, I think it drives the point home that allowing multiple calls to `Open()`is questionable at best. I think a lot less informative messages would have flied just fine if they wouldn't have happened to attract attention by their style. Let me use whatever literary style I want; I think this is another example of perfect consistency not being desirable.

---

Review status: 0 of 28 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/server/config.go, line 349 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUYKyTSYmxagrlbGxOL:-KUZ6vCKvLZSP6ervEC-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fa1da91fffe5d63db89c6d41373cc28fde50c64/pkg/server/config.go#L349)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this alignment is still broken.
</details>
Done.

---

*[pkg/server/config.go, line 383 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUYLB6byqzicPkEulVh:-KUYS2NdV6wxlnoGhQM0:b-xzks00) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fa1da91fffe5d63db89c6d41373cc28fde50c64/pkg/server/config.go#L383)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> instead of this boolean and a panic, you can reslice as i suggested.
</details>
the bool+panic serves to find double `Close()`

---

*[pkg/server/config.go, line 395 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUYLGieqVqcOwYDVoP3:-KUZDwNhyOttr_vJwNw0:b-rpswgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fa1da91fffe5d63db89c6d41373cc28fde50c64/pkg/server/config.go#L395)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> there is no need for this function to return a pointer, and the same is true of `NewEngines`. I believe `go vet` won't complain about `NewEngines`, but I am less sure about this one. You might need to work around it with:
> ```
> defer ep.engines = nil
> return NewEngines(ep.GetEngines())
> ```
> to appease `vet`.
</details>
ah yes, it's a recent change in `go vet`, right?
Switched over `NewEngine` to return by val and renamed. Did the same to `Move()`, but in the last commit, since until then calls to `Move()` are used in function calls that expect pointers.

---

*[pkg/server/config.go, line 402 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUYWhrP0I06Kbkoaymi:-KUZ5IvcQ1U10bQq7Jw5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fa1da91fffe5d63db89c6d41373cc28fde50c64/pkg/server/config.go#L402)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Nit: `s/AddEngine/Add/g`. You're already dealing with engines, the `Engine` suffix is repetitive. 
</details>
Done.

---

*[pkg/server/config.go, line 414 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUYL_7e-Vh1QRtKfa5s:-KUZ5eB80pLqI7zOH-x1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fa1da91fffe5d63db89c6d41373cc28fde50c64/pkg/server/config.go#L414)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> caller
</details>
Done.

---

*[pkg/server/config.go, line 419 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUYLcDBuXdbJ3FwX7bb:-KUZ5fBGsUBD9ucNYTvK:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fa1da91fffe5d63db89c6d41373cc28fde50c64/pkg/server/config.go#L419)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why do all of these error returns allocate?
</details>
done

---

*[server/context.go, line 350 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUXkKVXi_oVqx2K_5mJ:-KUYRIBrCnEx367iNVBj:b-34bjuw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b66495420766bb04574a99ebda09222c04a8bd65/server/context.go#L350)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> > The idea is that a client can construct one of these EnginesUniquePtr (now renamed) and immediately do defer e.Close(), and so if ownership hasn't been transferred by then, you're guaranteed to close them.
> > Not sure how the code here is different from what you're proposing.
> 
> The code you've written is just more complicated and tries to look like C++.
> 
> I don't understand the comment about the future commit. Can you squash whatever juggling was done here to one commit so that the same code doesn't need to be reviewed N times?
</details>
How is my code more complicated? It looks to me like your code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-:-KUYPk6gIqn2yDAD8aPK:b-7jme35)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
> For the first commit, "I hear it's not used" expressed uncertainty and invites discussion, if somebody thinks otherwise. It's hard to "be specific" when asserting that nobody is using something. I've added more specificity around the "racy" part though.

I'd prefer if commit messages served to replace discussion. These phrases don't only *express* uncertainty, they invite it. In this specific case, we don't use the thing you're talking about. The commit message is really only telling me that you didn't look into it enough to be able to be more assertive.

It's true that commit messages are written by humans for humans, but more succinctly, they are written by humans for strangers, and as a stranger I'd have a less than easy time trying to figure out that

> Is it possible to
Open() an engine multiple twice? Well, it was, for RocksDB, but that
didn't appear to be the intent of the interface. Is it OK to
Open/Close/Open? No.

really means

> The previous interface was ambiguous about whether an Engine
could be opened multiple times, a feature which was only used in
memory-based unit tests and was not properly supported by the
actual implementation. Remove the `Open` method from the
`Engine` interface so that an `Engine` is always open and this
subtlety removed.

Yes, that wouldn't win a Pulitzer, but last time I checked, it would win *my heart* and make sense to a wider audience, just like how people go to Wikipedia, not a poetry slam, for information.

> Let me use whatever literary style I want

This is a free country, just telling you that you're making it harder for others than it needs to be.

---

Review status: 0 of 28 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-:-KUaLXO2R2FMxO7G3yjJ:bezvjwa)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 1 of 26 files at r7, 1 of 4 files at r8, 1 of 6 files at r9, 1 of 3 files at r10, 20 of 26 files at r19, 2 of 4 files at r20, 2 of 6 files at r21, 4 of 4 files at r22.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/server/config.go, line 349 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUasTjkhgJSWr8xH0rj:-KUasTjkhgJSWr8xH0rk:b-kanu2l) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c7482c688c7a67573085152b8b929b67d1d2d1/pkg/server/config.go#L349)):*
> ```Go
> // The intended use is:
> //
> //	engines := MakeEngines(engines)
> ```

Using the name `engines` for both this new type and the slice that is passed in to it is confusing. If this new type is going to be `Engines`, then use `engineSlice` for the argument and `s/GetEngines/GetEngineSlice/`

---

*[pkg/server/config.go, line 384 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUat3ezBAM3bsgBHA7b:-KUat3ezBAM3bsgBHA7c:b-8rkc97) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c7482c688c7a67573085152b8b929b67d1d2d1/pkg/server/config.go#L384)):*
> ```Go
> }
> 
> // Close closes the engines. If Release() has been previously called, this is a
> ```

s/Release/Move/

---

*[pkg/server/config.go, line 398 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUavz2bOxeQh7vKXLlx:-KUavz2bOxeQh7vKXLly:bujjv4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c7482c688c7a67573085152b8b929b67d1d2d1/pkg/server/config.go#L398)):*
> ```Go
> 
> // Move transfers ownership of the engines to another Engines.
> // A future call to Close() on the original Engines will be a no-op.
> ```

This is not very idiomatic go (which is why this type is attracting so much bikeshedding). You don't even seem to be using this as far as I can tell except in CreateEngines. Unless you have more plans for this type, I think this whole type could be eliminated easily if you A) add a boolean in CreateEngine to indicate whether ownership of the engines has been passed to the caller so the defer knows whether it has to close them and B) add a static function `CloseEngines([]engine.Engine)` so the callers have a single function to defer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 28 of 29 files at r11, 7 of 26 files at r12, 1 of 4 files at r13.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/cli/start.go, line 264 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUYXlQxx5ntV--STO41:-KUYXlQxx5ntV--STO42:bd4dado) ([raw file](https://github.com/cockroachdb/cockroach/blob/9810f930de99667963eb8d66993fecaeb6148516/pkg/cli/start.go#L264)):*
> ```Go
> 				// Note that when dir is relative (as it is here) it is appended to the
> 				// engine's data directory.
> 				if err := e.Checkpoint(dir); err != nil {
> ```

this is also the only caller, right? why not remove this funciton altogether?

---

*[pkg/server/config.go, line 395 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUYLGieqVqcOwYDVoP3:-KUbKxyL7lvDAAoz0woO:b-itgny2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fa1da91fffe5d63db89c6d41373cc28fde50c64/pkg/server/config.go#L395)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> ah yes, it's a recent change in `go vet`, right?
> Switched over `NewEngine` to return by val and renamed. Did the same to `Move()`, but in the last commit, since until then calls to `Move()` are used in function calls that expect pointers.
</details>
Yep, this changed in vet for 1.7.

---

*[server/context.go, line 350 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUXkKVXi_oVqx2K_5mJ:-KUbL5dvsxVXufyQokgd:b-qmms0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/b66495420766bb04574a99ebda09222c04a8bd65/server/context.go#L350)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> How is my code more complicated? It looks to me like your code.
</details>
I probably didn't look past the first commit that introduced it. Like I said in another comment, having 5 commits that iterate on the same relatively simple type makes review very difficult and the history very messy. Consider if you can implement that type "correctly" in the first commit that needs it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
> I'd prefer if commit messages served to replace discussion. These phrases don't only express uncertainty, they invite it. In this specific case, we don't use the thing you're talking about. The commit message is really only telling me that you didn't look into it enough to be able to be more assertive.

The checkpointing in question is a "user feature". How would one exactly assert that nobody is using it? You are technically correct that I didn't look into it exhaustively - I have not talked to every engineer at Cockroach and every other contributor to ask them if they've ever used it. I don't intend to do that, either, so given this I can argue that the uncertainty is warranted. The commit also links to an issue where one can read more, although some level of uncertainty still remains. The statement was also supposed to be provocative in exactly this way - saying that I am not an expert in this feature, I am not 100% confident in what I'm doing, and inviting, as you say, discussion and opinions. Albeit the intended discussions was not supposed to be about this phrasing, so I guess the joke's on me.

For the second commit, I've like what Tobi says _less_. Besides amusing to my personal taste, my message was actually more informative. It gave more hints about what "not properly supported by the actual implementation" means. I don't think your message added preciseness, or any information really.
But I've thought about this a bit - maybe what's bothering you, possibly unconsciously, is that the commit message doesn't say anything about what the commit is doing to fix the issue; I think that'd be a reasonable complaint.

In any case, I'll try my very best to take this feedback to heart. I hold preciseness and clarity, where applicable, in high esteem. But I also want to say something else - I've lost sleep over this, and the same thing happened on previous discussions around commit messages. For me, personally, these disputes take a quite significant toll. For reasons I'm not entirely sure of, they take me a long time to process and respond to, and they otherwise suck away the joy of living. Maybe it's unfair, but it is a reality. So I'd kindly ask you to take this fact into consideration when deciding whether to flag something or not. I definitely don't want to silence any voices, and if you want to flag a message as subpar, please by all means, but please take into consideration whether the messages actually is subpar and what you believe to be the possible improvement is worth it.

---

Review status: 0 of 30 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/cli/start.go, line 264 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUYXlQxx5ntV--STO41:-KUhBC0Hhikt649kawHq:b-th2zke) ([raw file](https://github.com/cockroachdb/cockroach/blob/9810f930de99667963eb8d66993fecaeb6148516/pkg/cli/start.go#L264)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is also the only caller, right? why not remove this funciton altogether?
</details>
good call, done

---

*[pkg/server/config.go, line 398 at r22](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUavz2bOxeQh7vKXLlx:-KUhXyz7EltVHMG_Q3rV:b-czx031) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5c7482c688c7a67573085152b8b929b67d1d2d1/pkg/server/config.go#L398)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This is not very idiomatic go (which is why this type is attracting so much bikeshedding). You don't even seem to be using this as far as I can tell except in CreateEngines. Unless you have more plans for this type, I think this whole type could be eliminated easily if you A) add a boolean in CreateEngine to indicate whether ownership of the engines has been passed to the caller so the defer knows whether it has to close them and B) add a static function `CloseEngines([]engine.Engine)` so the callers have a single function to defer.
</details>
well so you and others have said that this is not idiomatic go, but doesn't that assume that there is an established idiom for what this struct is trying accomplish?I'm not aware of one. Probably because Go code, having a managed heap, doesn't deal as much as, say, C++, with object ownership concerns.
I was trying to propose such a pattern here. It does appear that it didn't quite win the hearts and minds, so I've now ripped it out; replaced with a simpler thing more similar to what you're suggesting.

FWIW, the intermediate commits in this PR used the `Move` functionality more than the final result. But at the end things were quite simplified, so I think it's not the right time for bikeshedding over such patterns.

---

*[server/context.go, line 350 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUXkKVXi_oVqx2K_5mJ:-KUhttXFTisSs0iZ67s4:b-ynslp9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b66495420766bb04574a99ebda09222c04a8bd65/server/context.go#L350)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I probably didn't look past the first commit that introduced it. Like I said in another comment, having 5 commits that iterate on the same relatively simple type makes review very difficult and the history very messy. Consider if you can implement that type "correctly" in the first commit that needs it.
</details>
it's gone

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-:-KUhBfUjsysdmGU8XKPr:bcks6r5)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> We can resurrect the checkpointing stuff if and when we need it in the future. 

---

Review status: 0 of 30 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-:-KUmDlp26l_uqRexsFSa:boukt8v)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
> The checkpointing in question is a "user feature". How would one exactly assert that nobody is using it?

That's almost a philosophical question, but for the purpose of the commit message all that matters is that you're sure beyond reasonable doubt, which I believe is the case here with everyone involved in its conception having agreed that it can go.

> For the second commit, I've like what Tobi says less.

Maybe it is objectively worse. The problem with code commentary (including commit messages) is that it doesn't matter whether I or you like one more than the other. What matters is what the future stranger who in two years' time has to figure out some code we touched makes of it. That's who I try to address (though it's hard to permanently suppress my innate urge to write everything in snarky rhyming couplets).

>  I've lost sleep over this, and the same thing happened on previous discussions around commit messages. For me, personally, these disputes take a quite significant toll. For reasons I'm not entirely sure of, they take me a long time to process and respond to, and they otherwise suck away the joy of living. 

That's certainly not the intention and I'm sorry about that. May I offer the hypothesis (from personal experience) that it takes a toll because you're taking them personally (took me a while not to, and it still happens occasionally).

I try to hold myself to a high standard (whether that works is an unrelated interesting question), and I think we share that interest and that you'd agree that honest criticism is the best way to get there. Endlessly pouring over commit messages may not, though, and I'll do my best to not make that as churny in the future.

---

Review status: 0 of 30 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-:-KUmL60nq3J5-gGZxhMS:biinhhc)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 29 files at r2, 2 of 30 files at r6, 28 of 29 files at r11, 2 of 30 files at r23, 20 of 26 files at r24, 2 of 4 files at r25, 2 of 6 files at r26, 4 of 4 files at r27.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 29 files at r2, 1 of 29 files at r11, 2 of 30 files at r23, 20 of 26 files at r24, 2 of 4 files at r25, 2 of 6 files at r26, 4 of 4 files at r27.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/server/config.go, line 436 at r27](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUmYefYZI_UgzUQzDT3:-KUmYefZhHUxZqx_qlhr:b-dabv2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/65cc68532c79180822ae2eb8a8f3fe6ff8ece1dd/pkg/server/config.go#L436)):*
> ```Go
> 		log.Infof(context.TODO(), "%d storage engines initialized", len(engines))
> 	}
> 	enginesCopy := engines
> ```

how come you don't want to encapsulate this in a method? seems easy to get this wrong.

---

*[pkg/server/server.go, line 533 at r27](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUmYyEPdSY3s7Gbs6Sk:-KUmYyEPdSY3s7Gbs6Sl:b2v0ryh) ([raw file](https://github.com/cockroachdb/cockroach/blob/65cc68532c79180822ae2eb8a8f3fe6ff8ece1dd/pkg/server/server.go#L533)):*
> ```Go
> 	engines, err := s.cfg.CreateEngines()
> 	if err != nil {
> 		return errors.Wrap(err, "failed to initialize stores")
> ```

"failed to create engines"?

---

*[pkg/server/server.go, line 536 at r27](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUmZ3JrZot3xlIWWjrN:-KUmZ3JrZot3xlIWWjrO:bw37ndg) ([raw file](https://github.com/cockroachdb/cockroach/blob/65cc68532c79180822ae2eb8a8f3fe6ff8ece1dd/pkg/server/server.go#L536)):*
> ```Go
> 	}
> 	s.stopper.AddCloser(&engines)
> 	s.engines = engines
> ```

why not do this directly above? there is already an `err` in scope.

---

*[pkg/storage/engine/in_mem.go, line 27 at r27](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUmZhdo_qItX93ST9lb:-KUmZhdo_qItX93ST9lc:b-s7d0pn) ([raw file](https://github.com/cockroachdb/cockroach/blob/65cc68532c79180822ae2eb8a8f3fe6ff8ece1dd/pkg/storage/engine/in_mem.go#L27)):*
> ```Go
> 
> // NewInMem allocates and returns a new, opened InMem engine.
> // The Engine returned needs to be Close()d.
> ```

To borrow language from the standard library: "The caller must call the engine's Close method when the engine is no longer needed."

---

*[pkg/storage/engine/rocksdb.go, line 282 at r27](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUm_6t0RydpZLMDavi_:-KUm_6t12Bx_lvn9EpUN:b-3iiyzl) ([raw file](https://github.com/cockroachdb/cockroach/blob/65cc68532c79180822ae2eb8a8f3fe6ff8ece1dd/pkg/storage/engine/rocksdb.go#L282)):*
> ```Go
> // doesn't yet exist at the specified directory, one is initialized
> // from scratch.
> // The caller must call Close() to close the database and releases resources.
> ```

s/releases/release/ but also use the standard library's language:
" The caller must call the engine's Close method when the engine is no longer needed."

---

*[util/nocopy.go, line 1 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUm_RRHhSj1ATV0ureD:-KUm_RRHhSj1ATV0ureE:be26wb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8733ef5cabdbe419a4dd85a942466128d5c618d/util/nocopy.go#L1)):*
> ```Go
> // Copyright 2016 The Cockroach Authors.
> ```

this file should be removed - looks like a bad rebase?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 30 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/server/config.go, line 436 at r27](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUmYefYZI_UgzUQzDT3:-KUsSdcazJBdJizaEmp4:b-2zsn7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/65cc68532c79180822ae2eb8a8f3fe6ff8ece1dd/pkg/server/config.go#L436)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how come you don't want to encapsulate this in a method? seems easy to get this wrong.
</details>
well I thought I had encapsulated it real nice with my uniquePtr and then nobody liked it :(
I'll leave the battle to be fought another day for a type that actually gets moved more than this engines.

---

*[pkg/server/server.go, line 533 at r27](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUmYyEPdSY3s7Gbs6Sk:-KUsTnXHoLdVNpw7rNrJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/65cc68532c79180822ae2eb8a8f3fe6ff8ece1dd/pkg/server/server.go#L533)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "failed to create engines"?
</details>
Done.

---

*[pkg/server/server.go, line 536 at r27](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUmZ3JrZot3xlIWWjrN:-KUsTo8lMIwzDnNCUqhd:b-4gyzqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/65cc68532c79180822ae2eb8a8f3fe6ff8ece1dd/pkg/server/server.go#L536)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why not do this directly above? there is already an `err` in scope.
</details>
because `s.Engines` was of type `[]Engine`, not `Engines`, and so I needed a temporary for taking the address of in the `AddCloser` call. I thought that makes sense, since the server does invoke `Close()` directly on `s.Engines`.
But I've changed it now.

---

*[pkg/storage/engine/in_mem.go, line 27 at r27](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUmZhdo_qItX93ST9lb:-KUsYhzh1UQFghN_qL6P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/65cc68532c79180822ae2eb8a8f3fe6ff8ece1dd/pkg/storage/engine/in_mem.go#L27)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> To borrow language from the standard library: "The caller must call the engine's Close method when the engine is no longer needed."
</details>
Done.

---

*[pkg/storage/engine/rocksdb.go, line 282 at r27](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUm_6t0RydpZLMDavi_:-KUsYvMLvLRQONiBUz1z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/65cc68532c79180822ae2eb8a8f3fe6ff8ece1dd/pkg/storage/engine/rocksdb.go#L282)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/releases/release/ but also use the standard library's language:
> " The caller must call the engine's Close method when the engine is no longer needed."
</details>
Done.

---

*[util/nocopy.go, line 1 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845#-KUm_RRHhSj1ATV0ureD:-KUsSU2Jm2C8RFl5ApS3:b-wu1pkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c8733ef5cabdbe419a4dd85a942466128d5c618d/util/nocopy.go#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this file should be removed - looks like a bad rebase?
</details>
was already gone

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 20 of 26 files at r29, 2 of 4 files at r30, 2 of 6 files at r31, 4 of 4 files at r32.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9845)*
<!-- Sent from Reviewable.io -->
